modal/volume.py

```diff
@@ -276,6 +277,28 @@ async def reload(self):
         try:
             await self._do_reload()
         except GRPCError as exc:
+            # TODO(staffan): This is very brittle and janky. We should return the open files from the server.
+            if exc.message == "there are open files preventing the operation":
+                # Attempt to identify what open files are problematic and include the path in the error message.
+                # This is best-effort and not necessarily bulletproof, as the view of open files inside the container
+                # might differ from that outside - but it will at least catch common errors.
+                vol_path = f"/__modal/volumes/{self.object_id}"
+
+                cwd = PurePosixPath(os.readlink("/proc/self/cwd"))
+                if cwd.is_relative_to(vol_path):
+                    raise RuntimeError(f"{exc.message}: the current working directory is inside the volume")
+
+                open_paths = []
+                for fd in os.listdir("/proc/self/fd"):
+                    try:
+                        path = PurePosixPath(os.readlink(f"/proc/self/fd/{fd}"))
+                        if path.is_relative_to(vol_path):
+                            rel_path = path.relative_to(vol_path)
+                            open_paths.append(rel_path.as_posix())
+                    except FileNotFoundError:
+                        pass
+                if open_paths:
+                    raise RuntimeError(f"{exc.message}: {open_paths} (relative to where the volume is mounted)")
```

Comment:
Will we just silently swallow this error if we can't work out the specific list of files? It's not totally obvious to me what's "brittle" or what the failure modes look like.

Reply:
no errors being swallowed - if we can't work out a list of files we will just raise `RuntimeError("there are open files preventing the operation")` error without any more info.

It's brittle / janky because there are things that can change server-side that will break this:
 * We rely on the specific error mesage (`"there are open files preventing the operation"`), which is :vomiting_face: :see_no_evil:
 * The specific path where volumes are mounted: `"/__modal/volumes/{self.object_id}"`

Where "break" just means reverting to the an error message will less info - so not the end of the world.

Ideally we'd figure out what files are causing issues server-side and just return them in the error message. Doing it client-side is a lot easier though.

Another reason to do that is that the container view of what files are open doesn't necessarily match what the host sees outside of the container, which is what really matters. The container view of open files will be whatever gVisor says it is - which may or may not be true :)

Reply:
Thanks that's helpful. Actually my question was fairly dumb anyway, I missed that the generic runtime error isn't in an else clause here, so we'll fall down to it if we don't exit on the highlighted lines.

I think the change here is a definite improvement, despite not being perfect. Thanks for adding it!